execution/executor/src/lib.rs

```diff
@@ -40,29 +42,227 @@ use std::{
+    committed_block_id: HashValue,
+}
+
+impl Drop for SpeculationBlock {
+    fn drop(&mut self) {
+        self.block_map
+            .lock()
+            .unwrap()
+            .remove(&self.id())
+            .expect("Speculation block must exist in block_map before being dropped.");
+        debug!("Speculation block {} is dropped.", self.id())
+    }
+}
+
+impl ExecutorCache {
+    fn new() -> Self {
+        Self {
+            synced_trees: ExecutedTrees::new_empty(),
+            committed_trees: ExecutedTrees::new_empty(),
+            heads: vec![],
+            block_map: Arc::new(Mutex::new(HashMap::new())),
+            committed_block_id: *PRE_GENESIS_BLOCK_ID,
+        }
+    }
+
+    fn new_with_startup_info(startup_info: StartupInfo) -> Self {
+        let mut cache = Self::new();
+        let ledger_info = startup_info.latest_ledger_info.ledger_info();
+        let committed_trees = ExecutedTrees::from(startup_info.committed_tree_state);
+        cache.update_block_tree_root(committed_trees, ledger_info);
+        if let Some(synced_tree_state) = startup_info.synced_tree_state {
+            cache.update_synced_trees(ExecutedTrees::from(synced_tree_state));
+        }
+        cache
+    }
+
+    fn committed_block_id(&self) -> HashValue {
+        self.committed_block_id
+    }
+
+    fn committed_trees(&self) -> &ExecutedTrees {
+        &self.committed_trees
+    }
+
+    fn synced_trees(&self) -> &ExecutedTrees {
+        &self.synced_trees
+    }
+
+    fn update_block_tree_root(
+        &mut self,
+        committed_trees: ExecutedTrees,
+        committed_ledger_info: &LedgerInfo,
+    ) {
+        let new_root_block_id = if committed_ledger_info.next_validator_set().is_some() {
+            // Update the root block id with reconfig virtual block id, to be consistent
+            // with the logic of Consensus.
+            let id = Block::<()>::make_genesis_block_from_ledger_info(committed_ledger_info).id();
+            debug!(
+                "Updated with a new root block {} as a virtual block of reconfiguration block {}",
+                id,
+                committed_ledger_info.consensus_block_id()
+            );
+            id
+        } else {
+            let id = committed_ledger_info.consensus_block_id();
+            debug!("updated with a new root block {}", id);
+            id
+        };
+        self.committed_block_id = new_root_block_id;
+        self.committed_trees = committed_trees.clone();
+        self.synced_trees = committed_trees;
+    }
+
+    fn update_synced_trees(&mut self, new_trees: ExecutedTrees) {
+        self.synced_trees = new_trees;
+    }
+
+    fn reset(&mut self) {
+        self.heads = vec![];
+        *self.block_map.lock().unwrap() = HashMap::new();
+    }
+
+    fn add_block(
+        &mut self,
+        parent_block_id: HashValue,
+        block: (
+            HashValue, /* block_id */
+            Vec<Transaction>,
+            ProcessedVMOutput,
```

Comment:
let's make this a type?

Reply:
This is used internally with only one callsite. IMO I don't think it is quite necessary and not suitable for this PR :)